src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -5,49 +5,132 @@
-            ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
-
-            compiler.evaluate(ast)
-            compiler.compile_additions
-            compiler.catalog_data_hash
+      def self.compile_ast(code, compile_options, boltlib_path)
+        # If the request requires that bolt be loaded we assume we are in a properly 
+        # configured PE environment
+        if compile_options.dig('options', 'compile_for_plan')
+          # CODEREVIEW: Where to we handle this error and what should it say?
+          unless boltlib_path
+            msg = "The path to boltlib modules must be provided in jruby-puppet boltlib-path puppetserver setting"
+            raise Puppet::Error, _(msg)
+          end
+
+          require 'bolt/apply_inventory'
+          require 'bolt/apply_target'
+          # TODO: Bolt's pal requires this so we dont have to do it in every boltib function that
+          # is explicitly forbidden in an apply block. Would it make sense to require in each function?
+          # or just do it here? 
+          require 'bolt/pal/issues'
+
+          Puppet[:node_name_value] = compile_options['certname']
+
+          env_conf = {
+            pre_modulepath: boltlib_path,
+            envpath: Puppet[:environmentpath],
+            facts: compile_options['facts']['values'],
+            variables: compile_options['variables']['values']
+          }
+
+          # Use the existing environment with the requested name
+          Puppet::Pal.in_environment(compile_options['environment'], env_conf) do |pal|
+            # TODO: Given we hide this from plan authors this current iteration has only
+            # the "required" data for now. Once we can get https://github.com/puppetlabs/bolt/pull/1770
+            # merged and promoted we can just use an empty hash.
+            fake_config = {
+              'transport' => 'redacted',
+              'transports' => {
+                'redacted' => 'redacted'
+              }
+            }
+            bolt_inv = Bolt::ApplyInventory.new(fake_config)
+            Puppet.override(bolt_inventory: bolt_inv) do
+              Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
+              # This compiler has been configured with a node containing
+              # the requested environment, facts, and variables, and is used
+              # to compile a catalog in that context from the supplied AST.
+              pal.with_catalog_compiler() do |compiler|
+                # TODO: PUP-10476 Explore setting these as default in PAL. They are the defaults in Puppet
+                Puppet[:strict] = :warning
+                Puppet[:strict_variables] = false
+
+                ast = build_program(code)
+                compiler.evaluate(ast)
+                compiler.evaluate_ast_node
+                compiler.compile_additions
+                compiler.catalog_data_hash
+              end
+            end
+          end
+        else
```

Comment:
Do we have a use case right now for this code branch? If not (and we can't think of one) I'm inclined to just get rid of it (and also the boolean in the request). As we pointed out early you, this endpoint isn't even documented.

Reply:
Not that I know of. I only added that code path to satisfy https://github.com/puppetlabs/puppetserver/pull/2329#discussion_r416055718

Reply:
Okay let's leave it then.